### ISSUE TITLE
why you implement DGC on pytorch

### ISSUE THREAD

**Issue by hbsun2113** [lusinlu/deep_gradient_compression#1]
As I know, Pytorch is ring all-reduce, I think DGC cannot help Pytorch to reduce the communication cost.

**Comment by hbsun2113** [comment]
In other words, what about your experiment? Can you tell me how much DGC improve the training speed?

**Comment by lusinlu** [comment]
Pytorch's all-reduce is not doing any sparsification, or compression of gradients, its just sending all the original gradients. And in case of DGC there is sparsity ~99.9%, without loss of accuracy  (as authors claim)

**Comment by hbsun2113** [comment]
SO why you implement DGC on Pytorch? I think it cannot help improve the training speed.

**Comment by lusinlu** [comment]
In distributed training, when you have physically separated machines, which is connected  through some network, you are not spending time on inference and backprop, but also on sending gradients through the network to centralized server, and amount of time that you will spend is depend on both network capacity and the size of the gradients, SO there is to approaches to speed up training on that stage - increase network capacity, or reduce size of the gradients. In case of DGC its working on reduction of gradient size. You can read more in details in the paper.

**Comment by hbsun2113** [comment]
I have read some papers like `Deep Gradient Compression`, but I think `gradient compression` is a benefit for `parameter server`, instead of `ring reduce`. And Pytorch is based on `ring reduce`, so I think it is useless to implement DGC on Pytorch.

**Comment by lusinlu** [comment]
I appreciate your opinion.Implementing or not, it depend on intentions.
